### PR TITLE
fix(lsp): lint warnings, default offset_encoding

### PR DESCRIPTION
**Pull request by glepnir**
- fix lint / analysis warnings
- locations_to_items(): get default offset_encoding from active client
- character_offset(): get default offset_encoding from active client

### CODE REVIEW

File: runtime/doc/lsp.txt [hunk 1]
```diff
@@ -1629,7 +1632,7 @@ apply_text_edits({text_edits}, {bufnr}, {offset_encoding})
     Parameters: ~
       • {text_edits}       (table) list of `TextEdit` objects
       • {bufnr}            (integer) Buffer id
-      • {offset_encoding}  (string) utf-8|utf-16|utf-32
+      • {offset_encoding}  (string|nil) utf-8|utf-16|utf-32|
```
Comment (justinmk) [inline runtime/doc/lsp.txt:1635]: why is the trailing `|` needed? seems weird

File: runtime/lua/vim/lsp/buf.lua [hunk 2]
```diff
@@ -697,7 +697,7 @@ local function on_code_action_results(results, ctx, options)
           return
         end
         apply_action(resolved_action, client)
-      end)
+      end, ctx.bufnr)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/buf.lua:700]: is this a bug fix?

File: runtime/lua/vim/lsp/protocol.lua [hunk 3]
```diff
@@ -898,7 +898,8 @@ function protocol.resolve_capabilities(server_capabilities)
   elseif type(textDocumentSync) == 'number' then
     -- Backwards compatibility
     if not TextDocumentSyncKind[textDocumentSync] then
-      return nil, 'Invalid server TextDocumentSyncKind for textDocumentSync'
+      vim.notify('Invalid server TextDocumentSyncKind for textDocumentSync', vim.log.levels.ERROR)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/protocol.lua:901]: was the returned message not being logged?

**Review by justinmk — APPROVED** [review]
This looks like a nice cleanup!

Nit: historically we tend to call non-functional cleanup simply "lint". "Reduce diagnostics" sounds like a change to the `diagnostics` subsystem.

File: runtime/doc/lsp.txt (hunk 1 above)
Comment (mfussenegger) [inline runtime/doc/lsp.txt:1635]: I think we should rather adapt the call-sites to provide the encoding instead of adding a `nil` case, otherwise the functions won't be working correctly.

File: runtime/lua/vim/lsp/codelens.lua [hunk 4]
```diff
@@ -53,9 +53,8 @@ local function execute_lens(lens, bufnr, client_id)
     return
   end
   client.request('workspace/executeCommand', command, function(...)
-    local result = vim.lsp.handlers['workspace/executeCommand'](...)
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/codelens.lua:56]: Afaik handlers can have results, this doesn't look right?

Comment (glepnir) [reply runtime/lua/vim/lsp/codelens.lua:56]: but this result never used  after this has `return result`

File: runtime/doc/lsp.txt (hunk 1 above)
Comment (glepnir) [reply runtime/doc/lsp.txt:1635]: right.  I will do it later.

File: runtime/lua/vim/lsp/buf.lua (hunk 2 above)
Comment (glepnir) [reply runtime/lua/vim/lsp/buf.lua:700]: yep. there is missing a param

File: runtime/lua/vim/lsp/protocol.lua (hunk 3 above)
Comment (glepnir) [reply runtime/lua/vim/lsp/protocol.lua:901]: idk. check this function usage . it not handle this error before so i am not sure it should be logged or not . just use notify through it

File: runtime/lua/vim/lsp/util.lua [hunk 5]
```diff
@@ -896,11 +896,12 @@ end
...
 --see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_applyEdit
 function M.apply_workspace_edit(workspace_edit, offset_encoding)
-  if offset_encoding == nil then
+  if not offset_encoding then
     vim.notify_once(
       'apply_workspace_edit must be called with valid offset encoding',
-      vim.log.levels.WARN
+      vim.log.levels.ERROR
     )
+    return
```
Comment (glepnir) [inline runtime/lua/vim/lsp/util.lua:904]: `apply_workspace_edit` most use in a callback  if there not pass a offset_encoding this function should not run right ? current buffer may not same as before. so i think can't set the first client `offset_encoding` in this function .  any thought?

Comment (mfussenegger) [reply runtime/lua/vim/lsp/util.lua:904]: It will currently fallback to `utf-16`, which is correct for almost all LSPs (clangd or rust-analyzer _can_ run with utf-8, but the neovim defaults to utf-16). I think turning this into an error with an early exit makes things worse.
We shouldn't do that before we haven't fixed all usages

(And it's general not nice to mix behavior changes into a generic cleanup PR focused on type annotation fixes)

File: runtime/lua/vim/lsp/util.lua [hunk 6]
```diff
@@ -1160,7 +1162,7 @@ end
 --- Shows document and optionally jumps to the location.
 ---
 ---@param location table (`Location`|`LocationLink`)
----@param offset_encoding "utf-8" | "utf-16" | "utf-32"
+---@param offset_encoding string utf-8|utf-16|utf-32
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:1165]: Can we leave this with the union type/enum syntax instead of generalizing it to string? Only the 3 values are allowed, not a string in general.

File: runtime/lua/vim/lsp/util.lua [hunk 7]
```diff
@@ -1173,6 +1175,7 @@ function M.show_document(location, offset_encoding, opts)
   end
   if offset_encoding == nil then
     vim.notify_once('show_document must be called with valid offset encoding', vim.log.levels.WARN)
+    return false
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:1178]: See comment further above. Please don't mix behavior changes like this into this PR

File: runtime/lua/vim/lsp/util.lua [hunk 8]
```diff
@@ -1226,6 +1229,7 @@ function M.jump_to_location(location, offset_encoding, reuse_win)
       'jump_to_location must be called with valid offset encoding',
       vim.log.levels.WARN
     )
+    offset_encoding = vim.lsp.get_active_clients({ bufnr = 0 })[1].offset_encoding
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:1232]: Same here, why the behavior change?

**Review by mfussenegger — CHANGES_REQUESTED** [review]

File: runtime/lua/vim/lsp/util.lua (hunk 8 above)
Comment (mfussenegger) [reply runtime/lua/vim/lsp/util.lua:1232]: > because they are too small just one line or two lines. it worth on a separator pr ?

Yes it's much easier to review

To unblock this PR: Could you revert all offset_encoding related changes in this PR? These need a closer look.  The rest looks good

Comment (glepnir) [reply runtime/lua/vim/lsp/util.lua:1232]: Okay and i see the `ctx.bufnr` you fix in another pr. need rebase then

Comment (glepnir) [reply runtime/lua/vim/lsp/util.lua]: @mfussenegger  LGTY?

File: runtime/doc/lsp.txt [hunk 9]
```diff
@@ -1697,14 +1700,15 @@ convert_signature_help_to_markdown_lines({signature_help}, {ft}, {triggers})
...
-      • {triggers}        optional list of trigger characters from the lsp
+      • {signature_help}  (table) Response of `textDocument/SignatureHelp`
+      • {ft}              (string|nil) filetype that will be use as the `lang`
+                          for the label markdown code block
+      • {triggers}        (table|nil) list of trigger characters from the lsp
                           server. used to better determine parameter offsets
 
     Return: ~
-        (list) of lines of converted markdown.
+        (table|nil) table list of lines of converted markdown.
```
Comment (justinmk) [inline runtime/doc/lsp.txt:1710]:
```suggestion
        (table|nil) list of lines of converted markdown.
```

File: runtime/doc/lsp.txt [hunk 10]
```diff
@@ -1732,15 +1736,15 @@ get_effective_tabstop({bufnr})          *vim.lsp.util.get_effective_tabstop()*
         (integer) indentation size
 
     See also: ~
-      • 'shiftwidth'
+      • shiftwidth
```
Comment (justinmk) [inline runtime/doc/lsp.txt:1739]: why was this changed? the single-quotes are needed to hyperlink in :help

File: runtime/doc/lsp.txt [hunk 11]
```diff
@@ -1767,11 +1772,11 @@ lookup_section({settings}, {section})          *vim.lsp.util.lookup_section()*
     Helper function to return nested values in language server settings
 
     Parameters: ~
-      • {settings}  a table of language server settings
-      • {section}   a string indicating the field of the settings table
+      • {settings}  (table) of language server settings
```
Comment (justinmk) [inline runtime/doc/lsp.txt:1775]:
```suggestion
      • {settings}  (table) language server settings
```

File: runtime/doc/lsp.txt (hunk 10 above)
Comment (glepnir) [reply runtime/doc/lsp.txt:1739]: a mistake forget restore.

File: runtime/doc/api.txt [hunk 12]
```diff
@@ -2015,7 +2015,7 @@ Buffer Functions                                                  *api-buffer*
 
 For more information on buffers, see |buffers|.
 
-Unloaded Buffers: ~
+Unloaded Buffers:~
```
Comment (justinmk) [inline runtime/doc/api.txt:2018]: revert

File: runtime/doc/lsp.txt [hunk 13]
```diff
@@ -1607,7 +1610,7 @@ convert_input_to_markdown_lines({input}, {contents})
                     lines. Defaults to {}.
 
     Return: ~
-        {contents}, extended with lines of converted markdown.
+        (table) extended with lines of converted markdown.
```
Comment (justinmk) [inline runtime/doc/lsp.txt:1613]:
```suggestion
        (table) {contents} extended with lines of converted markdown.
```

File: runtime/doc/lsp.txt (hunk 11 above)
Comment (justinmk) [reply runtime/doc/lsp.txt:1695]: ^ change wasn't pushed :)

File: runtime/doc/lsp.txt [hunk 14]
```diff
@@ -1853,10 +1858,11 @@ preview_location({location}, {opts})         *vim.lsp.util.preview_location()*
       definition)
 
     Parameters: ~
-      • {location}  a single `Location` or `LocationLink`
+      • {location}  (table) a single `Location` or `LocationLink`
 
     Return: ~
-        (bufnr,winnr) buffer and window number of floating window or nil
+        (integer|nil) buffer id of float window
+        (integer|nil) window id of float window
```
Comment (justinmk) [inline runtime/doc/lsp.txt:1865]: is there a way to express a tuple?

**Review by justinmk — COMMENTED** [review]
see comments

File: runtime/doc/lsp.txt (hunk 14 above)
Comment (glepnir) [reply runtime/doc/lsp.txt:1865]: this is generate from multiple return ,  tuple will like before. but it use `@retruns` I don't find `@returns` in luals annotations doc. so it might be wrong. 

```
---@return integer|nil buffer id of float window
---@return integer|nil window id of float window
```

**Comment by justinmk** [comment]
api docs need regen

**Review by mfussenegger — APPROVED** [review]

File: runtime/lua/vim/lsp/util.lua [hunk 15]
```diff
@@ -1828,13 +1833,15 @@ end)
...
+---                              default to first client of buffer
+---@return table list of items
 function M.locations_to_items(locations, offset_encoding)
   if offset_encoding == nil then
     vim.notify_once(
       'locations_to_items must be called with valid offset encoding',
       vim.log.levels.WARN
     )
+    offset_encoding = vim.lsp.get_active_clients({ bufnr = 0 })[1].offset_encoding
   end
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua:1845]: This reassigns the `offset_encoding` parameter but doesn't use the value. What is the purpose of that?

Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua:1845]: This reassigns `offset_encoding` which looks like a bug fix. In the future please:

1. call out these changes in the commit message and PR description
2. ideally, put them in a separate PR that isn't surrounded by 100s of lint changes

File: runtime/lua/vim/lsp/util.lua [hunk 16]
```diff
@@ -2167,15 +2174,16 @@ end
...
+---@return integer `offset_encoding` index of the character in line {row} column {col} in buffer {buf}
 function M.character_offset(buf, row, col, offset_encoding)
   local line = get_line(buf, row)
   if offset_encoding == nil then
     vim.notify_once(
       'character_offset must be called with valid offset encoding',
       vim.log.levels.WARN
     )
+    offset_encoding = vim.lsp.get_active_clients({ bufnr = buf })[1].offset_encoding
   end
```
Comment (justinmk) [inline runtime/lua/vim/lsp/util.lua:2187]: 👀

**Comment by jdrouhard** [comment]
The change to `vim.lsp.buf.clear_references({nil|bufnr})` was a breaking change for me. I had an autocmd like so:

```lua
      api.nvim_create_autocmd('CursorMoved', {
        group = au_group,
        buffer = bufnr,
        callback = vim.lsp.buf.clear_references,
        desc = 'vim.lsp.buf.clear_references',
      })
```

With this commit, the default table passed to autocmd callback functions no longer validates with the expected signature of `clear_references()` which is only nil or a bufnr. The corollary `buf.document_highlight()` still works since that one still doesn't take any params.

Not a big deal, but I had to do a bit of digging to figure out why my config broke when I updated my neovim checkout, and this commit wasn't marked as breaking.

@justinmk @glepnir

**Comment by kthenrique** [comment]
Same issue as described. What was your workaround @jdrouhard ?

**Comment by justinmk** [comment]
@glepnir is the `clear_references` change actually needed? changing behavior was not supposed to be part of this PR.

**Comment by glepnir** [comment]
behavior not changed.  because. there use `true ` in validate so nil can be used. so there is no need validate `nil`

**Comment by glepnir** [comment]
#24238
